src/lib.rs

```diff
@@ -118,6 +81,89 @@ where
     }
 }
 
+impl<T: Clone + IntoIterator> UpperHex for HexList<T>
+where
+    T::Item: AsRef<[u8]>,
+{
+    fn fmt(&self, f: &mut Formatter) -> Result {
+        let mut iter = self.0.clone().into_iter();
+        write!(f, "[")?;
+        if let Some(item) = iter.next() {
+            UpperHex::fmt(&HexFmt(item), f)?;
+        }
+        for item in iter {
+            write!(f, ", ")?;
+            UpperHex::fmt(&HexFmt(item), f)?;
+        }
+        write!(f, "]")
+    }
+}
+
+fn fmt(bytes: &[u8], f: &mut Formatter, case: Case) -> Result {
+    // TODO: Respect `f.width()`, `f.align()` and `f.fill()`.
+    let precision = f.precision().unwrap_or(DEFAULT_PRECISION);
+
+    // If the array is short enough, don't shorten it.
+    if 2 * bytes.len() <= precision {
+        for byte in bytes {
+            fmt_byte(f, *byte, case)?;
+        }
+        return Ok(());
+    }
+
+    // If the bytes don't fit and the ellipsis fills the maximum width, print only that.
+    if precision <= ELLIPSIS.len() {
+        return write!(f, "{:.*}", precision, ELLIPSIS);
+    }
+
+    // Compute the number of hex digits to display left and right of the ellipsis.
+    let num_hex_digits = precision.saturating_sub(ELLIPSIS.len());
+    let right = num_hex_digits / 2;
+    let left = num_hex_digits - right;
+
+    // Print the bytes on the left.
+    for byte in &bytes[..(left / 2)] {
+        fmt_byte(f, *byte, case)?;
+    }
+    // If odd, print only the first hex digit of the next byte.
+    if left & 1 == 1 {
+        fmt_digit(f, bytes[left / 2] >> 4, case)?;
+    }
+
+    // Print the ellipsis.
+    f.write_str(ELLIPSIS)?;
+
+    // If `right` is odd, print the second hex digit of a byte.
+    if right & 1 == 1 {
+        fmt_digit(f, bytes[(bytes.len() - right / 2 - 1)] & 0x0f, case)?;
+    }
+    // Print the remaining bytes on the right.
+    for byte in &bytes[(bytes.len() - right / 2)..] {
+        fmt_byte(f, *byte, case)?;
+    }
+    Ok(())
+}
+
+fn fmt_byte(f: &mut Formatter, byte: u8, case: Case) -> Result {
```

Comment:
This will cause a branch on every character printed. It's probably a minor thing, but I have seen this solved at compile time by making the uppercase/lowercase thing a type parameter (e.g. byteorder).

Something to consider.

Reply:
I don't know how the macro works, ideally, this match should be "resolvable" at compile time though. At the very least, enough `#[inline]` statements are needed to get this all the way to the call site where the string is defined!

Reply:
Good idea! I made `Case` a trait, so it can be statically dispatched. I also added `#[inline]` everywhere, although I'm not convinced it is necessary, and haven't looked at the difference in the assembly.

Reply:
Well, it would take a bit of work to see it, as it's only relevant when calling cross-crate!